cli/src/maintenance.rs

```diff
@@ -1356,6 +1376,43 @@ impl SolidoState {
             Some(self_slice_start_slot)
         }
     }
+
+    /// Return None if we observe we moved past `1 -
+    /// SolidoState::END_OF_EPOCH_THRESHOLD`%. Return Some(()) if the above
+    /// condition fails or `self.stake_unstake_any_time` is set to
+    /// `true`.
+    pub fn confirm_should_stake_unstake_in_current_slot(&self) -> Option<()> {
+        match self.stake_time {
+            StakeTime::Anytime => Some(()),
+            StakeTime::OnlyNearEpochEnd => {
+                // Get the slot that the current epoch started.
+                let slots_epoch_begin = self
+                    .epoch_schedule
+                    .get_first_slot_in_epoch(self.clock.epoch);
+                let next_epoch_begin = self
+                    .epoch_schedule
+                    .get_first_slot_in_epoch(self.clock.epoch + 1);
+                let slots_per_epoch = next_epoch_begin
+                    .checked_sub(slots_epoch_begin)
+                    .expect("Next epoch's slot should be always greater than previous.");
+                let slot_past_epoch = self.clock.slot.checked_sub(slots_epoch_begin).expect(
+        "Current slot is less than the beginning of the epoch's slot. This shouldn't happen.",
+    );
+                let ratio = Rational {
+                    numerator: slots_per_epoch.checked_sub(slot_past_epoch).expect(
```

Comment:
Aah, I now undertand where my confusion about the tests came from, this function is wrong after all ... We should stake late in the epoch, but due to the subtract here, that is when `ratio` is close to zero. So the comparison below should be `ratio < SolidoState::END_OF_EPOCH_THRESHOLD`, not `>`. That’s why all the expected results of the tests are also flipped; the function should return `None` when it’s not yet time to stake (so the `?` in calls to `confirm_should_stake_unstake_in_current_slot()?` makes the caller return immediately).

One way to fix would be to swap flip the `>` to a `<`, but I think the end result will be simpler if we change the definition of `END_OF_EPOCH_THRESHOLD` to 19/20, change the definition of `ratio` here to just `slot_past_epoch/slots_per_epoch`, and leave the comparison as `ratio > END_OF_EPOCH_THRESHOLD`.

Reply:
Yeah, that's right 😬. I changed it, should be good now.